src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -334,6 +334,12 @@ RubyBinding binding(Frame callerFrame, Object self, Object[] rubyArgs, RootCallT
                 @Cached(
                         value = "getAdoptedNode(this).getEncapsulatingSourceSection()",
                         allowUncached = true, neverDefault = false) SourceSection sourceSection) {
+            final InternalMethod method = RubyArguments.tryGetMethod(callerFrame);
+            if (method == null || method.getDeclaringModule().toString().equals("Truffle::CExt")) {
```

Comment:
For reference it seems `tryGetMethod` is looking up the ruby method that is the context of the frame calling `binding`.
So in the case of the test
https://github.com/oracle/truffleruby/blob/0c3f184c762ed0b5cce42d5b33f8c1b8bcb46658/spec/ruby/optional/capi/binding_spec.rb#L14
is backed by
https://github.com/oracle/truffleruby/blob/0c3f184c762ed0b5cce42d5b33f8c1b8bcb46658/spec/ruby/optional/capi/ext/binding_spec.c#L8-L9
and the method of the caller frame then points to
https://github.com/oracle/truffleruby/blob/0c3f184c762ed0b5cce42d5b33f8c1b8bcb46658/lib/truffle/truffle/cext.rb#L1071

Reply:
There's probably also a better check here than doing a string comparison on the method's module, but we hadn't found anything.
Any pointers on what else we might be able to use for this?

Reply:
There is `org.truffleruby.core.CoreLibrary#truffleCExtModule` and then it can be compared with `==`.

But comparing on the module is hacky, because it means every `Truffle::CExt` method can no longer access `binding`.
Which seems a problem notably for:
https://github.com/oracle/truffleruby/blob/610cea3302a7302c5ec3efd1a6f89b26467d04c3/lib/truffle/truffle/cext.rb#L2039-L2041

That can be worked around, but it shows it's rather hacky.

I'm not sure what's a good solution for detecting C ext methods, maybe/probably we should define `Truffle::CExt#rb_define_method` in Java eventually. Or find another way to mark those.

But I think for now a more useful check would be to check if `sourceSection == CoreLibrary.JAVA_CORE_SOURCE_SECTION`, which means a method defined in Java.
Getting a binding from there will always be empty, which seems not useful, so that seems the corresponding case to CRuby having methods defined in C and for which no meaningful binding can be given.

An example test for that is:
```
$ ruby -e 'p method(:binding).call'
-e:1:in `binding': Cannot create Binding object for non-Ruby caller (RuntimeError)
```
Currently on TruffleRuby it's:
```
$ ruby -e 'p method(:binding).call'
#<Binding:0xc8>
$ ruby -e 'p method(:binding).call.local_variables'
[]
```

Reply:
BTW for the existing spec in spec/ruby/optional/capi/binding_spec.rb I would think we already raise an error for that case, because the callerFrame should be `null` in that case since it is a call from LLVM/Sulong.
But maybe there is something in between which means there is a frame somehow.

Reply:
On the master branch this spec fails with
```
CApiBindingSpecs Kernel#binding raises when called from C FAILED
Expected RuntimeError but no exception was raised (#<Binding:0x368> was returned)
```

I loaded a debugger to check the sourceSection and it is not the same as `CoreLibrary.JAVA_CORE_SOURCE_SECTION`.

This is what I see:
![image](https://github.com/oracle/truffleruby/assets/142719/ac76d5b0-34a9-41f5-b8b0-edf8f8a469fe)

If I use a different ruby script that calls binding from ruby code the sourceSection here doesn't have any different properties from the above (other than the `path`) that I can see 🤔

Reply:
Right, so the SourceSection there is https://github.com/oracle/truffleruby/blob/677ac08b296b086a7d6ede73f6cbebe7c8027a6a/lib/truffle/truffle/cext.rb#L1060

So let's not try to pass this specific spec for now, because it would require changing `rb_funcallv` which is out of scope.

But instead let's add another spec, using `method(:binding).call`, and that should raise the RuntimeError, and that one we can use the check for JAVA_CORE_SOURCE_SECTION.

Reply:
One thing worth trying to pass that C API spec would be to change `SendWithoutCExtLockBaseNode` to call `callWithFrameAndBlock` without a `frame` (so with `null` instead). But that may have other side effects which might be undesirable.

Reply:
^ but the frame was added back in 6fbb9eeece3f643db183e4687818da15e4711497 probably for 796df1651f6d0d66839336813a5ad414ece462d0.
And 9c0f7ce2a2b2d83d04919bca240f05de4b0624f5 seems also related.
So I think best to not touch that for now.

Reply:
https://github.com/oracle/truffleruby/pull/3449 Is an alternative implementation that uses something similar to the commit you pointed to 9c0f7ce